src/main/scala/firrtl/analyses/ConnectionGraph.scala

```diff
@@ -0,0 +1,684 @@
+    * Top/a:A>out         List(Top>a.in)
+    * Top>a.out           List()
+    * Top>out             List()
+    * when we reach Top/a:A>in the stack is List
+    */
+  private val portConnectivityStack: mutable.HashMap[ReferenceTarget, List[ReferenceTarget]] =
+    mutable.HashMap.empty[ReferenceTarget, List[ReferenceTarget]]
+
+  /** Records connectivities found while BFS is executing, from a module's source port to sink ports of a module
+    *
+    * All keys and values are local references.
+    *
+    * A BFS search will first query this map. If the query fails, then it continues and populates the map. If the query
+    * succeeds, then the BFS shortcuts with the values provided by the query.
+    *
+    * Because this BFS implementation uses a priority queue which prioritizes exploring deeper instances first, a
+    * successful query during BFS will only occur after all paths which leave the module from that reference have
+    * already been searched.
+    */
+  private val bfsShortCuts: mutable.HashMap[ReferenceTarget, mutable.HashSet[ReferenceTarget]] =
+    mutable.HashMap.empty[ReferenceTarget, mutable.HashSet[ReferenceTarget]]
+
+  /** Records connectivities found after BFS is completed, from a module's source port to sink ports of a module
+    *
+    * All keys and values are local references.
+    *
+    * If its keys contain a reference, then the value will be complete, in that all paths from the reference out of
+    * the module will have been explored
+    *
+    * For example, if Top>in connects to Top>out1 and Top>out2, then foundShortCuts(Top>in) will contain
+    * Set(Top>out1, Top>out2), not Set(Top>out1) or Set(Top>out2)
+    */
+  private val foundShortCuts: mutable.HashMap[ReferenceTarget, mutable.HashSet[ReferenceTarget]] =
+    mutable.HashMap.empty[ReferenceTarget, mutable.HashSet[ReferenceTarget]]
+
+  /** Returns whether a previous BFS search has found a shortcut out of a module, starting from target
+    *
+    * @param target
+    * @return
+    */
+  def hasShortCut(target: ReferenceTarget): Boolean = getShortCut(target).nonEmpty
+
+  /** Optionally returns the shortcut a previous BFS search may have found out of a module, starting from target
+    *
+    * @param target
+    * @return
+    */
+  def getShortCut(target: ReferenceTarget): Option[Set[ReferenceTarget]] =
+    foundShortCuts.get(target.pathlessTarget).map(set => set.map(_.setPathTarget(target.pathTarget)).toSet)
+
+  /** Returns the shortcut a previous BFS search may have found out of a module, starting from target
+    *
+    * @param target
+    * @return
+    */
+  def shortCut(target: ReferenceTarget): Set[ReferenceTarget] = getShortCut(target).get
+
+  /** Returns a new, reversed connection graph where edges point from sinks to sources
+    *
+    * @return
+    */
+  def reverseConnectionGraph: ConnectionGraph = new ConnectionGraph(circuit, digraph.reverse, irLookup)
+
+  def getTag(node: ReferenceTarget,
```

Comment:
Should we add scaladoc for this? It seems like all of the tagging stuff is not used in this PR. I believe it was necessary for the clock domain tracking stuff, which has been stripped away. Personally, I think this code is not particularly generic for the connectivity graph, and I remember trying to refactor it anyways in a different branch. How do you feel about removing it for now? This includes `getTag`, `setTag`, `tagPath`, etc.

Reply:
Removing it :)